### PR TITLE
Az999 get fsm backpressure

### PR DESCRIPTION
**Pull request by reiddraper**
I've refactored the GET fsm to have the caller _ask_ for chunks instead of receiving them in the process mailbox. This adds implicit backpressure and makes writing the WM code easier because we don't have to deal with the potential variability in the WM proc pid. The tests have been updated to reflect this. Take a look at the tests to see the new usage.

### CODE REVIEW

File: src/riak_moss_get_fsm.erl [hunk 1]
```diff
@@ -46,6 +54,9 @@
                 bucket :: term(),
                 key :: term(),
                 value_cache :: binary(), 
+                metadata_cache :: term(), 
+                chunk_queue :: term(),
+                reply_pid :: pid(),
```
Comment (argv0) [inline src/riak_moss_get_fsm.erl]: reply_pid seems to be unused?

File: src/riak_moss_get_fsm.erl [hunk 2]
```diff
@@ -63,24 +74,32 @@ stop(Pid) ->
...
+
 %% ====================================================================
 %% gen_fsm callbacks
 %% ====================================================================
 
-init([From, Bucket, Key]) ->
-    State = #state{from=From, bucket=Bucket, key=Key,
-                   get_module=riak_moss_riakc},
+init([Bucket, Key]) ->
+    Queue = queue:new(),
```
Comment (argv0) [inline src/riak_moss_get_fsm.erl:89]: not an issue for this PR, but we might want to consider a variant of riak_repl's bounded_queue in the future.

File: src/riak_moss_get_fsm.erl [hunk 3]
```diff
@@ -92,6 +111,10 @@ prepare(timeout, #state{bucket=Bucket, key=Key, get_module=GetModule}=State) ->
     spawn_link(?MODULE, normal_retriever, [self(), GetModule, Bucket, Key]),
```
Comment (argv0) [inline src/riak_moss_get_fsm.erl]: we probably want to stash this pid and match on it in handle_info - under heavy load pid reuse can become an issue.

File: src/riak_moss_get_fsm.erl [hunk 4]
```diff
@@ -63,24 +74,32 @@ stop(Pid) ->
 continue(Pid) ->
     gen_fsm:send_event(Pid, continue).
 
+get_metadata(Pid) ->
+    gen_fsm:sync_send_event(Pid, get_metadata).
+
+get_next_chunk(Pid) ->
+    gen_fsm:sync_send_event(Pid, get_next_chunk).
```
Comment (argv0) [inline src/riak_moss_get_fsm.erl:82]: consider moving riak_moss_put_fsm:send_sync_event to a utility module and using that for get_metadata/get_next_chunk

File: src/riak_moss_get_fsm.erl (hunk 1 above)
Comment (reiddraper) [inline src/riak_moss_get_fsm.erl]: fixed in 2d92fba424a4223995bf630aa864ad9a2ca26c1e

File: src/riak_moss_get_fsm.erl (hunk 2 above)
Comment (reiddraper) [inline src/riak_moss_get_fsm.erl:89]: agreed.

File: src/riak_moss_get_fsm.erl (hunk 3 above)
Comment (reiddraper) [inline src/riak_moss_get_fsm.erl]: fixed in fe870137e56ec1e010c39a25a1c63fc047116df2

**Comment by argv0** [comment]
+1 to merge

**Pull request merged**